test_util/src/pty.rs

```diff
@@ -0,0 +1,437 @@
+    pub fn new(
+      program: impl AsRef<Path>,
+      args: &[&str],
+      cwd: &str,
+      maybe_env_vars: Option<HashMap<String, String>>,
+    ) -> Self {
+      // https://docs.microsoft.com/en-us/windows/console/creating-a-pseudoconsole-session
+      unsafe {
+        let mut size: COORD = std::mem::zeroed();
+        size.X = 800;
+        size.Y = 500;
+        let mut console_handle = std::ptr::null_mut();
+        let (stdin_read_handle, stdin_write_handle) = create_pipe();
+        let (stdout_read_handle, stdout_write_handle) = create_pipe();
+
+        let result = CreatePseudoConsole(
+          size,
+          stdin_read_handle.as_raw(),
+          stdout_write_handle.as_raw(),
+          0,
+          &mut console_handle,
+        );
+        assert_eq!(result, S_OK);
+
+        let mut environment_vars = maybe_env_vars.map(get_env_vars);
+        let mut attribute_list = ProcThreadAttributeList::new(console_handle);
+        let mut startup_info: STARTUPINFOEXW = std::mem::zeroed();
+        startup_info.StartupInfo.cb =
+          std::mem::size_of::<STARTUPINFOEXW>() as u32;
+        startup_info.lpAttributeList = attribute_list.as_mut_ptr();
+
+        let mut proc_info: PROCESS_INFORMATION = std::mem::zeroed();
+        let command = format!(
+          "\"{}\" {}",
+          program.as_ref().to_string_lossy(),
+          args.join(" ")
+        )
+        .trim()
+        .to_string();
+        let mut application_str =
+          to_windows_str(&program.as_ref().to_string_lossy());
+        let mut command_str = to_windows_str(&command);
+        let mut cwd = to_windows_str(cwd);
+
+        assert_win_success!(CreateProcessW(
+          application_str.as_mut_ptr(),
+          command_str.as_mut_ptr(),
+          ptr::null_mut(),
+          ptr::null_mut(),
+          FALSE,
+          EXTENDED_STARTUPINFO_PRESENT | CREATE_UNICODE_ENVIRONMENT,
+          environment_vars
+            .as_mut()
+            .map(|v| v.as_mut_ptr() as LPVOID)
+            .unwrap_or(ptr::null_mut()),
+          cwd.as_mut_ptr(),
+          &mut startup_info.StartupInfo,
+          &mut proc_info,
+        ));
+
+        // close the handles that the pseudoconsole now has
+        drop(stdin_read_handle);
+        drop(stdout_write_handle);
+
+        // start a thread that will close the pseudoconsole on process exit
+        let thread_handle = WinHandle::new(proc_info.hThread);
+        std::thread::spawn({
+          let thread_handle = thread_handle.duplicate();
+          let console_handle = WinHandle::new(console_handle);
+          move || {
+            WaitForSingleObject(thread_handle.as_raw(), INFINITE);
+            // wait for the reading thread to catch up
+            std::thread::sleep(Duration::from_millis(200));
+            // close the console handle which will close the
+            // stdout pipe for the reader
+            ClosePseudoConsole(console_handle.take());
+          }
+        });
```

Comment:
The winapi examples create a separate thread to do the reading: https://github.com/microsoft/terminal/blob/fcb42b262f0bab726184edccc42e9632dcfba452/samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp#L37 and the documentation [on ending a session](https://docs.microsoft.com/en-us/windows/console/creating-a-pseudoconsole-session#ending-the-pseudoconsole-session).

I didn't want to do this though because it would complicate the `Pty` trait. Instead, this code creates a thread that waits for the pseudoconsole process to exit and when it does it closes the stdout pipe which will unblock the reader.